src/resolver.js

```diff
@@ -1,164 +1,70 @@
+ *
+ * Returns the value or a link and the partial mising path. This way the
+ * IPLD Resolver can fetch the link and continue to resolve.
+ *
+ * @param {Buffer} binaryBlob - Binary representation of a PB block
+ * @param {string} [path='/'] - Path that should be resolved
+ * @returns {Object} result - Result of the path it it was resolved successfully
+ * @returns {*} result.value - Value the path resolves to
+ * @returns {string} result.remainderPath - If the path resolves half-way to a
+ *   link, then the `remainderPath` is the part after the link that can be used
+ *   for further resolving.
  */
-exports.resolve = (binaryBlob, path, callback) => {
-  waterfall([
-    (cb) => util.deserialize(binaryBlob, cb),
-    (node, cb) => {
-      // Return the deserialized block if no path is given
-      if (!path) {
-        return callback(null, {
-          value: node,
-          remainderPath: ''
-        })
-      }
-
-      const split = path.split('/')
-
-      if (split[0] === 'Links') {
-        let remainderPath = ''
-
-        // all links
-        if (!split[1]) {
-          return cb(null, {
-            value: node.links.map((l) => l.toJSON()),
-            remainderPath: ''
-          })
-        }
-
-        // select one link
-
-        const values = {}
-
-        // populate both index number and name to enable both cases
-        // for the resolver
-        node.links.forEach((l, i) => {
-          const link = l.toJSON()
-          values[i] = values[link.name] = {
-            cid: link.cid,
-            name: link.name,
-            size: link.size
-          }
-        })
-
-        let value = values[split[1]]
-
-        // if remainderPath exists, value needs to be CID
-        if (split[2] === 'Hash') {
-          value = { '/': value.cid }
-        } else if (split[2] === 'Tsize') {
-          value = value.size
-        } else if (split[2] === 'Name') {
-          value = value.name
-        }
-
-        remainderPath = split.slice(3).join('/')
-
-        cb(null, { value: value, remainderPath: remainderPath })
-      } else if (split[0] === 'Data') {
-        cb(null, { value: node.data, remainderPath: '' })
-      } else {
-        // If split[0] is not 'Data' or 'Links' then we might be trying to refer
-        // to a named link from the Links array. This is because go-ipfs and
-        // js-ipfs have historically supported the ability to do
-        // `ipfs dag get CID/a` where a is a named link in a dag-pb.
-        const values = {}
-
-        node.links.forEach((l, i) => {
-          const link = l.toJSON()
-          values[link.name] = {
-            cid: link.cid,
-            name: link.name,
-            size: link.size
-          }
-        })
-
-        const value = values[split[0]]
-
-        if (value) {
-          return cb(null, {
-            value: { '/': value.cid },
-            remainderPath: split.slice(1).join('/')
-          })
-        }
+exports.resolve = async (binaryBlob, path) => {
```

Comment:
`async` can be removed from this, it'll just make an unnecessary `Promise`

Reply:
It's for consistency with other formats. I'd like to keep it for now until it turns out to be an issue.

Reply:
I get the consistency argument but really think the format API needs to be "MaybePromise" rather than "Promise", or just make `resolve()` completely synchronous. None of the formats do anything async within `resolve()`, they're all the same pattern as this one (other than raw which is even simpler):

* https://github.com/ipld/js-ipld-dag-cbor/pull/107/files#diff-fe783c782beb75b01586f9afb052da1bR21
* https://github.com/ipld/js-ipld-zcash/pull/39/files#diff-fe783c782beb75b01586f9afb052da1bR21
* https://github.com/ipld/js-ipld-bitcoin/pull/48/files#diff-fe783c782beb75b01586f9afb052da1bR21
* https://github.com/ipld/js-ipld-git/pull/51/files#diff-fe783c782beb75b01586f9afb052da1bR21
* https://github.com/ipld/js-ipld-raw/pull/32/files#diff-1fdf421c05c1140f6d71444ea2b27638R22

So the `async` here is theoretical at best. I like that it might be accommodating a potential of encryption or something else novel, but so far it doesn't, and that could be accomplished by stating in the format specification that it "may be a Promise" so users may want to opt for `await` if they want to be maximally compatible. I suspect most uses of the current code are sticking to specific implementations though and that flexibility can be left to the next generation stack.

As I understand, Matteo's performance analysis identified as the faux-async with `setImmediate()` everywhere was the biggest cost for js-ipfs. Replacing them with `async` doesn't buy much at all, it's just different sugar for deferring something that isn't actually asynchronous.

Reply:
I'll make `resolve()` sync for the same reason that I made `serialize()/deserialize()` sync. Thanks for this outside perspective. I'm currently too deep into this without seeing the bigger picture.